File: src/cod_bringup/scripts/odom_to_tf.py
#!/usr/bin/env python3
"""
Converts nav_msgs/Odometry messages into odom→base_link TF broadcasts.
Used in Gazebo simulation where the bridge provides odometry but no TF.

Uses raw=True subscription to work around the ROS2 Humble + ros_gz_bridge
deserialization bug (RuntimeError: Unable to convert call argument to Python object).
"""

import rclpy
from rclpy.node import Node
from rclpy.serialization import deserialize_message
from nav_msgs.msg import Odometry
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped


class OdomToTf(Node):
    def __init__(self):
        super().__init__('odom_to_tf')

        self.declare_parameter('odom_topic', 'chassis_odometry_gt')
        self.declare_parameter('odom_frame', 'odom')
        self.declare_parameter('base_frame', 'base_link')

        odom_topic = self.get_parameter('odom_topic').value
        self.odom_frame = self.get_parameter('odom_frame').value
        self.base_frame = self.get_parameter('base_frame').value

        self.tf_broadcaster = TransformBroadcaster(self)
        # raw=True: receive CDR bytes to avoid C++ deserialization bug with
        # ros_gz_bridge Odometry messages on ROS2 Humble
        self.sub = self.create_subscription(
            Odometry, odom_topic, self.odom_cb, 10, raw=True)

        self.get_logger().info(
            f'Publishing TF [{self.odom_frame}] -> [{self.base_frame}] '
            f'from topic [{odom_topic}]')

    def odom_cb(self, raw_msg):
        msg = deserialize_message(raw_msg, Odometry)
        t = TransformStamped()
        t.header.stamp = msg.header.stamp
        t.header.frame_id = self.odom_frame
        t.child_frame_id = self.base_frame
        t.transform.translation.x = msg.pose.pose.position.x
        t.transform.translation.y = msg.pose.pose.position.y
        t.transform.translation.z = msg.pose.pose.position.z
        t.transform.rotation = msg.pose.pose.orientation
        self.tf_broadcaster.sendTransform(t)


def main(args=None):
    rclpy.init(args=args)
    node = OdomToTf()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/fake_vel_transform/src/odom_to_tf.cpp
// Converts nav_msgs/Odometry to odom→base_link TF.
// Used in Gazebo simulation where the bridge provides odometry but no TF.

#include "fake_vel_transform/odom_to_tf.hpp"

#include <geometry_msgs/msg/transform_stamped.hpp>

namespace fake_vel_transform
{

OdomToTf::OdomToTf(const rclcpp::NodeOptions & options)
: Node("odom_to_tf", options)
{
  this->declare_parameter<std::string>("odom_topic", "chassis_odometry_gt");
  this->declare_parameter<std::string>("odom_frame", "odom");
  this->declare_parameter<std::string>("base_frame", "base_link");

  std::string odom_topic;
  this->get_parameter("odom_topic", odom_topic);
  this->get_parameter("odom_frame", odom_frame_);
  this->get_parameter("base_frame", base_frame_);

  tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
    odom_topic, 10,
    std::bind(&OdomToTf::odomCallback, this, std::placeholders::_1));

  RCLCPP_INFO(
    get_logger(), "Publishing TF [%s] -> [%s] from topic [%s]",
    odom_frame_.c_str(), base_frame_.c_str(), odom_topic.c_str());
}

void OdomToTf::odomCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
{
  geometry_msgs::msg::TransformStamped t;
  t.header.stamp = msg->header.stamp;
  t.header.frame_id = odom_frame_;
  t.child_frame_id = base_frame_;
  t.transform.translation.x = msg->pose.pose.position.x;
  t.transform.translation.y = msg->pose.pose.position.y;
  t.transform.translation.z = msg->pose.pose.position.z;
  t.transform.rotation = msg->pose.pose.orientation;
  tf_broadcaster_->sendTransform(t);
}

}  // namespace fake_vel_transform

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(fake_vel_transform::OdomToTf)
